share/eds/store.go

```diff
@@ -0,0 +1,207 @@
+	}
+
+	s := &EDSStore{
+		basepath: basepath,
+		dgstr:    dagStore,
+		topIdx:   invertedRepo,
+		carIdx:   fsRepo,
+		mounts:   r,
+	}
+
+	s.bs, err = NewEDSBlockstore(s)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create EDSBlockstore: %w", err)
+	}
+
+	return s, nil
+}
+
+func (s *EDSStore) Start(ctx context.Context) error {
+	return s.dgstr.Start(ctx)
+}
+
+func (s *EDSStore) Stop() error {
+	return s.dgstr.Close()
+}
+
+// Put stores the given data square with DataRoot's hash as a key.
+//
+// The square is verified on the Exchange level, and Put only stores the square, trusting it.
+// The resulting file stores all the shares and NMT Merkle Proofs of the EDS.
+// Additionally, the file gets indexed s.t. store.Blockstore can access them.
+func (s *EDSStore) Put(ctx context.Context, root share.Root, square *rsmt2d.ExtendedDataSquare) error {
+	key := root.String()
+	f, err := os.OpenFile(s.basepath+blocksPath+key, os.O_CREATE|os.O_WRONLY, 0600)
+	if err != nil {
+		return err
+	}
+
+	err = WriteEDS(ctx, square, f)
+	if err != nil {
+		return err
+	}
+
+	ch := make(chan dagstore.ShardResult, 1)
+	err = s.dgstr.RegisterShard(ctx, shard.KeyFromString(key), &mount.FSMount{
+		FS:   os.DirFS(s.basepath + blocksPath),
+		Path: key,
+	}, ch, dagstore.RegisterOpts{})
+	if err != nil {
+		return err
+	}
+
+	result := <-ch
+	return result.Error
+}
+
+// GetCAR takes a DataRoot and returns a buffered reader to the respective EDS serialized as a CARv1 file.
+//
+// The Reader strictly reads the CAR header and first quadrant (1/4) of the EDS, omitting all the NMT Merkle proofs.
+// Integrity of the store data is not verified.
+//
+// Caller must Close returned reader after reading.
+func (s *EDSStore) GetCAR(ctx context.Context, root share.Root) (io.ReadCloser, error) {
+	key := root.String()
+
+	ch := make(chan dagstore.ShardResult, 1)
+	err := s.dgstr.AcquireShard(ctx, shard.KeyFromString(key), ch, dagstore.AcquireOpts{})
+	if err != nil {
+		return nil, err
+	}
+
+	result := <-ch
+	if result.Error != nil {
+		return nil, result.Error
+	}
+	return result.Accessor, nil
+}
+
+// Blockstore returns an IPFS Blockstore providing access to individual shares/nodes of all EDS
+// registered on the Store. NOTE: The Blockstore does not store whole Celestia Blocks but IPFS blocks.
+// We represent `shares` and NMT Merkle proofs as IPFS blocks and IPLD nodes so Bitswap can access those.
+func (s *EDSStore) Blockstore() blockstore.Blockstore {
+	return s.bs
+}
+
+// Remove removes EDS from Store by the given share.Root and cleans up all the indexing.
+func (s *EDSStore) Remove(ctx context.Context, root share.Root) error {
+	key := root.String()
+
+	ch := make(chan dagstore.ShardResult, 1)
+	err := s.dgstr.DestroyShard(ctx, shard.KeyFromString(key), ch, dagstore.DestroyOpts{})
+	if err != nil {
+		return err
+	}
+
+	result := <-ch
+	if result.Error != nil {
+		return result.Error
+	}
+
+	dropped, err := s.carIdx.DropFullIndex(shard.KeyFromString(key))
```

Comment:
I checked the DAGStores's logic and they really don't drop the full index on Destroy. Good Catch. We need to add this to the ADR.

Reply:
👍🏼 lets update the storage ADR with this + EDSStore name change + share.Root stuff once networking ADR is completed

Reply:
#1382